src/renderers/shared/fiber/ReactFiberBeginWork.js

```diff
@@ -103,34 +122,40 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
         }
       }
     }
+
     instance.props = props;
     var nextChildren = instance.render();
     reconcileChildren(current, workInProgress, nextChildren);
-    workInProgress.pendingWorkPriority = NoWork;
-    return workInProgress.childInProgress;
+
+    return workInProgress.child;
   }
 
   function updateHostComponent(current, workInProgress) {
-    var nextChildren = workInProgress.pendingProps.children;
+    const nextChildren = workInProgress.pendingProps.children;
+    if (workInProgress.pendingProps.hidden &&
+        workInProgress.pendingWorkPriority !== OffscreenPriority) {
+      // If this host component is hidden, we can bail out on the children.
+      // We'll rerender the children later at the lower priority.
+
+      // It is unfortunate that we have to do the reconciliation of these
+      // children already since that will add them to the tree even though
+      // they are not actually done yet. If this is a large set it is also
+      // confusing that this takes time to do right now instead of later.
 
-    let priority = workInProgress.pendingWorkPriority;
-    if (workInProgress.pendingProps.hidden && priority !== OffscreenPriority) {
-      // If this host component is hidden, we can reconcile its children at
-      // the lowest priority and bail out from this particular pass. Unless, we're
-      // currently reconciling the lowest priority.
-      // If we have a child in progress already, we reconcile against that set
-      // to retain any work within it. We'll recreate any component that was in
-      // the current set and next set but not in the previous in progress set.
-      // TODO: This attaches a node that hasn't completed rendering so it
-      // becomes part of the render tree, even though it never completed. Its
-      // `output` property is unpredictable because of it.
+      if (workInProgress.progressedPriority === OffscreenPriority) {
+        // If we already made some progress on the offscreen priority before,
+        // then we should continue from where we left off.
+        workInProgress.child = workInProgress.progressedChild;
```

Comment:
Maybe we should change `reconcileChildrenAtPriority` to take this as an argument so we don't need to temporarily set it like this.